### ISSUE TITLE
Akamai and EDNS Client Subnet issues

### ISSUE THREAD

**Issue by trmdi** [AdguardTeam/AdGuardDNS#729]
### Issue Details

    $ CLASS=CH RRTYPE="TXT" dnslookup id.server tls://dns.adguard.com                                      
    dnslookup v1.9.1                                        
    dnslookup result (elapsed 276.594529ms):
    ;; opcode: QUERY, status: NXDOMAIN, id: 60293           
    ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, 
    AUTHORITY: 1, ADDITIONAL: 6                                            
    ;; QUESTION SECTION:                                    
    ;id.server.     CH       TXT                            ;; 
    AUTHORITY SECTION:
    .       1237    IN      SOA     a.root-servers.net. 
    nstld.verisign-grs.com. 2024012202 1800 900 
    604800 86400    
    ;; ADDITIONAL SECTION:
    client-ip.adguard-dns.com.      3600    CH      TXT    "14.226.99.x"                                          
    server-ip.adguard-dns.com.      3600    CH      TXT    "94.140.14.14"
    country.adguard-dns.com.        3600    CH      TXT    "VN"
    asn.adguard-dns.com.    3600    CH      TXT     "45899"
    subdivision.adguard-dns.com.    3600    CH      TXT    "HP"                                                     resp.res- 
    type.adguard-dns.com.  3600    CH      TXT    "normal"

### Expected Behavior

When you are in the North of Vietnam, your ECS should be set to ones in the North, because the client should connect to CDNs in Hanoi/HCM/Hongkong for shorter distances and latency, instead of Singapore.

### Actual Behavior

    $ RRTYPE=TXT dnslookup o-o.myaddr.l.google.com tls://dns.adguard.com
    dnslookup v1.9.1                                        
    dnslookup result (elapsed 285.909138ms):
    ;; opcode: QUERY, status: NOERROR, id: 18760            
    ;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

    ;; QUESTION SECTION:                                    ;o- 
    o.myaddr.l.google.com.       IN       TXT

    ;; ANSWER SECTION:
    o-o.myaddr.l.google.com.        60      IN      TXT    "edns0-client-subnet 222.255.238.0/24"
    o-o.myaddr.l.google.com.        60      IN      TXT    "138.199.46.250"

    $ ipinfo.sh 222.255.238.0           [11:14:28]
    {
    "ip": "222.255.238.0",
    "hostname": "static.vnpt.vn",
    "city": "Ho Chi Minh City",
    "region": "Ho Chi Minh",
    "country": "VN",
    "loc": "10.8230,106.6296",
    "org": "AS45899 VNPT Corp",
    "postal": "71606",
    "timezone": "Asia/Ho_Chi_Minh"
    }

**Comment by trmdi** [comment]
From a screenshot in the video below, I see subdivision is applied to CN, IN... but not to VN. Is it right?
https://github.com/AdguardTeam/AdGuardDNS/issues/659#issuecomment-1904324548

**Comment by trmdi** [comment]
$ dig starling-oversea.byteoversea.com @dns.adguard.com +short                                         starling-oversea.byteoversea.com.ttdns2.com.
starling-oversea.byteoversea.com.edgekey.net.           e11942.a.akamaiedge.net.
23.215.7.32                                             23.215.7.28
23.215.7.19                                             23.215.7.16
23.215.7.22                                             23.215.7.23
23.215.7.20                                             23.215.7.25
23.215.7.18                                             

$ dig starling-oversea.byteoversea.com @dns.google +short                                              starling-oversea.byteoversea.com.ttdns2.com.
starling-oversea.byteoversea.com.edgekey.net.           e11942.a.akamaiedge.net.
113.171.10.152
113.171.10.144
113.171.10.162
113.171.10.160
113.171.10.155
113.171.10.154
113.171.10.146
113.171.10.139                                          113.171.10.145

$ ping -c3 23.215.7.15              [11:49:13] PING 23.215.7.15 (23.215.7.15) 56(84) bytes of data.
64 bytes from 23.215.7.15: icmp_seq=1 ttl=57 time=47.5 ms
64 bytes from 23.215.7.15: icmp_seq=2 ttl=57 time=46.1 ms
64 bytes from 23.215.7.15: icmp_seq=3 ttl=57 time=48.9 ms

--- 23.215.7.15 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2003ms
rtt min/avg/max/mdev = 46.087/47.489/48.892/1.145 ms

$ ping -c3 113.171.10.155           [11:49:33]
PING 113.171.10.155 (113.171.10.155) 56(84) bytes of data.
64 bytes from 113.171.10.155: icmp_seq=1 ttl=58 time=6.05 ms
64 bytes from 113.171.10.155: icmp_seq=2 ttl=58 time=4.58 ms
64 bytes from 113.171.10.155: icmp_seq=3 ttl=58 time=3.22 ms

--- 113.171.10.155 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2003ms
rtt min/avg/max/mdev = 3.220/4.618/6.052/1.156 ms

**Comment by ameshkov** [comment]
Right, the subdivision logic at this moment is enabled for 4 countries only: US, Russia, China and India.

Interesting data, thank you! We'll experiment with it a bit, I am not yet sure that it should be solved. We should first figure out why Akamai returns Singapore addresses for this subnet, it's unclear from the BGP information what's so different with this prefix compared to other prefixes announced by AS45899.

**Comment by trmdi** [comment]
> We should first figure out why Akamai returns Singapore addresses for this subnet

Because that subnet belongs to HCM which is near Singapore? I guess a subnet in HN should be better.
![R9iiRwD](https://github.com/AdguardTeam/AdGuardDNS/assets/11857205/bf1b1195-23c8-4c58-b426-4f3a1292e028)

**Comment by ameshkov** [comment]
@trmdi HCM?

As I see both subnets (yours and the one that AG DNS uses for ECS) belongs to the same AS45899.

* ECS - `222.255.238.0/24`
* Your subnet - `14.226.99.0/24`

**Comment by ameshkov** [comment]
Ah okay I got it, HCM = Ho Chi Minh

**Comment by trmdi** [comment]
Sorry, HCM means Ho Chi Minh city. (In the south of Vietnam)

**Comment by ameshkov** [comment]
Maybe you're right, we should check how Akamai nameserver responses change depending on what location the subnet belongs to.

Internal task ID: AGDNS-1881

**Comment by ameshkov** [comment]
Just got one more idea. Vietnam has quite a lot of different subdivisions in MaxMind (it uses ISO 3166-2). According to the ISO 3166-2 standard, Vietnam is divided into [58 provinces and 5 municipalities](https://en.wikipedia.org/wiki/ISO_3166-2:VN), but for network routing it is probably not that important and it'd be enough to do a "north-south" split.

@trmdi does that sound reasonable to you? If so, how would you split the subdivisions between north/south?

I think we could probably do the same for large countries as well, split subdivisions between west/east/north/south/central in order to improve cache efficiency.

**Comment by trmdi** [comment]
Yes, I think splitting between west/east/north/south/central is fine.
Specifically, Vietnam area is S-shaped, narrow and stretched from the north to the south, so I think splitting to north/south is enough for network routing.

**Comment by ameshkov** [comment]
Okay it's not that easy.

Something's strange with Akamai. They indicate in responses that they don't rely on ECS:
```
dig -t a starling-oversea.byteoversea.com. @8.8.8.8 +subnet=14.226.99.0/24

...
CLIENT-SUBNET: 14.226.99.0/24/0
...
```

Now one more strange thing.

There's another domain that I tested with: `store.akamai.steamstatic.com.`

For some reason via Google DNS Akamai responds very accurately with Cyprus IP addresses.

What's strange is that when I pass a custom subnet via dig to Google DNS, it continues to respond with accurate Cyprus IP addresses. I honestly do not understand what's going on there and how on earth could Akamai learn about my location considering that I am being routed to Google DNS servers located in Switzerland (and Google DNS does not have any servers in Cyprus anyways).

**Comment by ameshkov** [comment]
Here's what I suspect:

1. In the case of Akamai nameservers, Google DNS sends my real subnet in ECS even if I specify the subnet in the query.
2. Akamai only accepts ECS from a select list of DNS servers. They accept it from Google DNS and OpenDNS, but not from us or Quad9.

**Comment by vttale** [comment]
Your suspicious are both correct.    They will only respond to ECS from known resolvers with whom they have a contractual relationship, and relationship they have with Akamai does not permit the resolvers to forward ECS from clients.

**Comment by ameshkov** [comment]
@vttale thank you for confirming my suspicion!

On a side note, this looks like a weird business decision for Akamai. They should've been interested in speeding up connections for all Internet users, not for a part of them.

**Comment by ameshkov** [comment]
Found it:
https://community.akamai.com/customers/s/article/What-is-the-ECS-onboarding-agreement-for-No-Client-Side-ECS

![image](https://github.com/AdguardTeam/AdGuardDNS/assets/5947035/866fc936-dd73-4c7e-a5b4-3270c2a19e7d)

Well, our ECS implementation cannot fulfil what they're requiring. I'll try to contact them and maybe find some middle ground.

**Comment by vttale** [comment]
> On a side note, this looks like a weird business decision for Akamai. They should've been interested in speeding up connections for all Internet users, not for a part of them.

From their point of view, their mapping is proprietary information based on extensive measurement of the Internet.   They don't want that trove of data to be easily walked through iterative ECS queries.

**Comment by trmdi** [comment]
Today I noticed that Akamai has accepted the ECS sent by AdGuard DNS. Is that correct? It's a big improvement.
If so, could you please consider using two different ECS IPs for North and South Vietnam? That would be a further optimization.
(Current situation: I'm in the North, Adguard DNS returns CDNs in the South while there are CDNs in the North with much better pings and distances, because of the wrong ECS)

```bash
$ dig +subnet=14.226.99.0/24 cf.shopee.vn.akamaized.net @dns.adguard-dns.com +short
a1958.w16.akamai.net.
113.171.234.27
113.171.234.9
$ dig +subnet=2.4.0.0/24 cf.shopee.vn.akamaized.net @dns.adguard-dns.com +short
a1958.w16.akamai.net.
23.222.250.100
23.222.250.75
```

**Comment by pictosun** [comment]
> Today I noticed that Akamai has accepted the ECS sent by AdGuard DNS. Is that correct? It's a big improvement. If so, could you please consider using two different ECS IPs for North and South Vietnam? That would be a further optimization. (Current situation: I'm in the North, Adguard DNS returns CDNs in the South while there are CDNs in the North with much better pings and distances, because of the wrong ECS)

Could this also mean, that the issues from https://github.com/AdguardTeam/AdGuardDNS/issues/777 are solved?

**Comment by ameshkov** [comment]
There is a possibility. I haven't yet got the confirmation from Akamai that they started accepting ECS **everywhere**, but there's indeed a chance that it will normalize now

**Comment by instacompvtltd** [comment]
We are seeing the same issue on our ISP (INSTACOM, Pakistan, downstream of Transworld). Android devices using Private DNS (dns.adguard.com) experience TikTok videos stalling, while the same devices work normally with our ISP DNS, Google DNS, or Cloudflare. Our investigation shows AdGuard DNS appears to map our subscribers to different Akamai/TikTok cache nodes than other resolvers. dig +subnet testing indicates Google returns our local CDN cache, while AdGuard returns different cache IPs. This appears to be a CDN/ECS mapping issue rather than a DNS resolution problem. We're happy to provide packet captures, dig outputs, ASN, and prefix information if it helps investigate.

root@instadns:~# dig +subnet=103.131.212.0/24 cf.shopee.vn.akamaized.net @dns.adguard-dns.com +short
a1958.w16.akamai.net.
59.103.92.224
59.103.92.227
root@instadns:~# dig +subnet=103.131.213.0/24 cf.shopee.vn.akamaized.net @dns.adguard-dns.com +short
a1958.w16.akamai.net.
59.103.92.224
59.103.92.227
root@instadns:~# dig +subnet=103.131.214.0/24 cf.shopee.vn.akamaized.net @dns.adguard-dns.com +short
a1958.w16.akamai.net.
59.103.92.227
59.103.92.224
root@instadns:~# dig +subnet=103.131.215.0/24 cf.shopee.vn.akamaized.net @dns.adguard-dns.com +short
a1958.w16.akamai.net.
59.103.92.227
59.103.92.234

We are expecting our upstream clustor ips:

110.93.x.x which is Transworld Cluster.

**Comment by trmdi** [comment]
> We are seeing the same issue on our ISP (INSTACOM, Pakistan, downstream of Transworld). Android devices using Private DNS (dns.adguard.com) experience TikTok videos stalling, while the same devices work normally with our ISP DNS, Google DNS, or Cloudflare.
> ...
> We are expecting our upstream clustor ips:
> 
> 110.93.x.x which is Transworld Cluster.

I think ECS is anonymized but quite accurate in your case. Is `103.131.213.0/24` your IP?

```bash
# check ECS
test:~/ $ dig TXT whoami.ds.akahelp.com @dns.adguard-dns.com +short +subnet=103.131.213.0/24
"ip" "119.156.228.203"
"ecs" "119.156.228.0/24/24"
"ns" "79.127.170.12"

# specified ECS
test:~/ $ ipinfo.sh 103.131.213.0
{
  "ip": "103.131.213.0",
  "city": "Multan",
  "region": "Punjab",
  "country": "PK",
  "loc": "30.1968,71.4782",
  "org": "AS138368 INSTACOM Pvt. LTD",
  "postal": "60700",
  "timezone": "Asia/Karachi"
}

# anonymized ECS by Adguard DNS 
test:~/ $ ipinfo.sh 119.156.228.0
{
  "ip": "119.156.228.0",
  "city": "Kot Addu",
  "region": "Punjab",
  "country": "PK",
  "loc": "30.4691,70.9670",
  "org": "AS138368 INSTACOM Pvt. LTD",
  "postal": "34050",
  "timezone": "Asia/Karachi"
}

# Adguard DNS server at my location
test:~/ $ ipinfo.sh 79.127.170.12
{
  "ip": "79.127.170.12",
  "hostname": "unn-79-127-170-12.datapacket.com",
  "city": "Singapore",
  "region": "Singapore",
  "country": "SG",
  "loc": "1.2897,103.8501",
  "org": "AS60068 Datacamp Limited",
  "postal": "018989",
  "timezone": "Asia/Singapore"
}
```

**Comment by instacompvtltd** [comment]
We have following subnets:

103.131.212.0/24
103.131.213.0/24
103.131.214.0/24
103.131.215.0/24
119.156.228.0/24

Actually following Akamai Nodes has no advertisement of our prefixes. Thats why Tiktok stuck and do not run on our netwok while using your dns.adguard.com

59.103.92.227
59.103.92.234

We need our clusters to be provided by dns.adguard.com. Any solution for it?

**Comment by ameshkov** [comment]
We'll see what can be done about that, thank you!

**Comment by instacompvtltd** [comment]
Ok we are wating with thanks.

**Comment by instacompvtltd** [comment]
Please update on issue. Our users are suffering from bad experience on tiktok videos.

**Comment by trmdi** [comment]
Today I rechecked the original issue and noticed a change. It's still the same ECS but the location has changed from the South to the North. It's an unexpected change that is good for the North people but not for the South ones.
Not sure how you chose the subnet for the ECS, and how accurate it is?

<details>
<summary>Current data</summary>

    test:~/ $ dig TXT whoami.ds.akahelp.net @dns.adguard-dns.com +tls -4 +short
    "ecs" "222.255.238.0/24/24"
    "ns" "89.222.116.129"
    "ip" "222.255.238.19"
    test:~/ $ ipinfo.sh 222.255.238.0
    {
      "ip": "222.255.238.0",
      "hostname": "static.vnpt.vn",
      "city": "Yên Bái",
      "region": "Lao Cai",
      "country": "VN",
      "loc": "21.7229,104.9113",
      "org": "AS45899 VNPT Corp",
      "postal": "33150",
      "timezone": "Asia/Bangkok"
    }
</details>

<details>
<summary>The original data</summary>

    $ ipinfo.sh 222.255.238.0
    {
    "ip": "222.255.238.0",
    "hostname": "static.vnpt.vn",
    "city": "Ho Chi Minh City",
    "region": "Ho Chi Minh",
    "country": "VN",
    "loc": "10.8230,106.6296",
    "org": "AS45899 VNPT Corp",
    "postal": "71606",
    "timezone": "Asia/Ho_Chi_Minh"
    }
</details>

**Comment by EugeneOne1** [comment]
Hello @instacompvtltd and apologies for the late response.  We've added more variety to the ECS selection for Pakistan. Could you please check if the responses are any different now?
